### ISSUE TITLE
Sound Visualization becomes increasingly delayed

### ISSUE THREAD

**Issue by JonasWechsler** [psieg/Lightpack#243]
I use S/PDIF for audio out, and in the dropdown there are four S/PDIF output options- two labeled "Digital Audio (S/PDIF) (High Definition Audio Device) (loopback)" and two without "(loopback)". Only one of these produces any color change, which is one of the loopback options. 

The color is initially very tightly linked to the music, but after a few minutes there's a long (30 seconds or more) delay and the colors need to "play out" before any settings changes occur, e.g. if I change to mood lamp it can take a couple minutes while the color blink away.

**Comment by zomfg** [comment]
Which version?
Default visualizer?
How long is few minutes?
Does this happen in screen grabbing mode (with desktop duplication grabber) if you set grab interval to 25ms or less and play a video for the same amount of time?

**Comment by PauliusLe** [comment]
Same issue for me with 5.11.2.19. It looks like sound visualization is producing more data than it can be sent to LED controller and this data stays in some kind of buffer. Delay becomes visible pretty quickly, after 10s visualization is 1s behind. When music stops, visualization continues for some time. It happened for me with old 5.11.2.17 version. But at that time I was using 240 LEDs and I thought that data rate is too big for arduino nano. Currently I am using 104 LEDs, no issues with 5.11.2.17 version. Screen grabbing also works fine in 5.11.2.19.

**Comment by zomfg** [comment]
> Screen grabbing also works fine in 5.11.2.19

To confirm, set it to 25ms or slightly less and force it to reach max framerate (with a video for example) long enough to see (or not) the delay.

**Comment by JonasWechsler** [comment]
Yes, I get a similar delay with 20ms delay if I'm playing a video with lots of color changing (I played [this video](https://youtu.be/cgk31hjfB0E) for 10-20 seconds). I'm using a 115200 baud rate. With normal usage (not a lot of colors changing), it updates normally.

The delay increases with time in sound visualizer mode, like PauliusLe said. I ran the sound visualizer for 30 seconds and then switch to mood lamp, and it took 14 or 15 seconds to update.

**Comment by zomfg** [comment]
Can you try setting something like 500k baud rate?

**Comment by JonasWechsler** [comment]
That seems to fix it. I'm running 138 LEDs- not sure if that's a lot. I'll run it for a while to see if I get the same problem, but for now it gets out of sync much slower, if at all.

**Comment by zomfg** [comment]
After a quick test with a slower delay, upping the baud rate is my recommended option. With just higher delay it looks pretty "numb".

Use [this awesome post](https://www.partsnotincluded.com/ambilight/calculating-adalight-framerate-limits/) to eyeball your baud rate depending on your needs. (In my experience I was able to get slightly more frames before artifacts)

**Comment by Fabulizer** [comment]
Hello,
I have the same problem as Jonas. I have 127 LEDs. Screen grabbing works just fine with baud rate 115200 and grab interval 50ms, but sound visualization gets really delayed just after few seconds. 
I've tried switching baud rate, but it doesn't work then at all. I've changed the code and reconfigured Prismatik, there is no errors, but it still doesn't work.
I think if there was some way to limit the amount of data sent to Arduino it would work.
Somebody has any solution or knows why can't I change baud rate? I really would like to have sound visualization working.

**Comment by JonasWechsler** [comment]
When you say it doesn't work with a higher baud rate, is it completely
dark? You probably just need to change your frequency in the firmware for
the arduino, so that it can receive the higher refresh rate.

**Comment by Fabulizer** [comment]
> When you say it doesn't work with a higher baud rate, is it completely dark? You probably just need to change your frequency in the firmware for the arduino, so that it can receive the higher refresh rate.

If you mean to change code #define serialRate 115200 to #define serialRate 921600, I did it.

**Comment by zomfg** [comment]
Did you go directly to 900k? not every chip will be able to do it. Try 500k

**Comment by Fabulizer** [comment]
I've also tried 460800 and same result - black.
Maybe it's worth mension that I got Arduino Nano with old bootloader, but idk.

**Comment by zomfg** [comment]
Try other rates (lower/higher), and make sure you have the same rate on both sides
and see [this](https://github.com/psieg/Lightpack/issues/243#issuecomment-456596527) if you haven't already.
